File: src/core/awakening/actions.ts
import {
  type Controller,
} from "../view";
import { firstResponderFor } from "../responder";
import { ownedDescendants } from "../view/ownership";

const ACTION_BOUND = Symbol.for("framework.actionBound");

interface ActionAnnotatedElement extends HTMLElement {
  [ACTION_BOUND]?: Set<string>;
}

// DOM wiring for target-action attributes. Dispatch starts at
// `firstResponderFor(sender)` and then walks the responder chain.
export function wireActionsInto(c: Controller, host: HTMLElement = c.view.element): () => void {
  const disposers: Array<() => void> = [];
  const els = ownedDescendants(host, {
    includeRoot: true,
    match: (el) => el.hasAttribute("data-action"),
  });
  els.forEach((el) => {
    const spec = el.getAttribute("data-action")!.trim();
    if (!spec) return;
    const annotated = el as ActionAnnotatedElement;
    const bound = (annotated[ACTION_BOUND] ??= new Set<string>());
    if (bound.has(spec)) return;
    bound.add(spec);
    const [maybeEvent, maybeName] = spec.includes(":") ? spec.split(":") : [undefined, spec];
    const eventName = (maybeEvent ?? "click").trim();
    const actionName = (maybeName ?? "").trim();
    if (!actionName) return;
    const handler = (event: Event): void => {
      (firstResponderFor(el) ?? c).performAction(actionName, el, event);
    };
    el.addEventListener(eventName, handler);
    disposers.push(() => {
      el.removeEventListener(eventName, handler);
      bound.delete(spec);
    });
  });
  return () => {
    for (const dispose of disposers) dispose();
  };
}

File: opal/lib/swill/actions.rb
# backtick_javascript: true

module Swill
  module Actions
    module_function

    def wire(controller, root = controller.view.element)
      Bindings.owned_elements(controller, root, "[data-action]").each do |element|
        next if `#{element}.__swill_action__`

        specification = `#{element}.getAttribute("data-action")`.to_s.strip
        next if specification.empty?

        event_name, action_name = parse(specification)
        listener = lambda do |event|
          controller.perform_action(action_name, element, event)
        end

        `#{element}.__swill_action__ = true`
        `#{element}.addEventListener(#{event_name}, #{listener})`
      end
    end

    def parse(specification)
      return specification.split(":", 2) if specification.include?(":")

      ["click", specification]
    end
  end
end
